### DISCUSSION TITLE
LSP-like functionality in the LazyVim config/plugin files?

### DISCUSSION THREAD

**Discussion by fredrikaverpil** [LazyVim/LazyVim#862] [Q&A] [answered]
Is it possible to somehow achieve LSP-like completions and have `gd` and friends work when editing the LazyVim config files?

Like for example, when I add a `require("xxx").setup({...})`, can I somehow make it possible to easily gotodef on `xxx` and hit `<S-k>` on the `setup` function to inspect its available arguments?

I'm new to Neovim/LazyVim/Lua and I feel I'm missing out on something here. I'm aware of neodev, but not sure if this should cover this case. My LazyVim config is located [here](https://github.com/fredrikaverpil/dotfiles/tree/main/nvim-lazyvim/lua).

## Solution

Turns out that I had symlinked the `.lua` plugin files into place, and it seems this made the workspace act up. Adding a `.neoconf.json` to my symlink source directory (dotfiles repo) solved it:

```json
{
  "neodev": {
    "library": {
      "enabled": true,
      "plugins": true
    }
  },
  "neoconf": {
    "plugins": {
      "lua_ls": {
        "enabled": true
      }
    }
  }
}
```

**Comment by fredrikaverpil** [comment]
<details>
<summary>Bad take (edited out)</summary>




Doesn't seem like you can add `gd` support for e.g. `vim.*` (doesn't seem like neovim stores this as readable lua files) but at least one can get completions to work.

I had to extend the lspconfig with this, now I can get completion in all my neovim/LazyVim config lua files (taken from the lspconfig docs [here](https://github.com/neovim/nvim-lspconfig/blob/master/doc/server_configurations.md#lua_ls)).

```lua
-- extend the LazyVim lspconfig
  {
    "neovim/nvim-lspconfig",
    config = function()
      local lsp_config = require("lspconfig")
      -- lua language server setup
      lsp_config.lua_ls.setup({
        settings = {
          Lua = {
            runtime = {
              -- Tell the language server which version of Lua you're using (most likely LuaJIT in the case of Neovim)
              version = "LuaJIT",
            },
            diagnostics = {
              -- Get the language server to recognize the `vim` global
              globals = { "vim" },
            },
            workspace = {
              -- Make the server aware of Neovim runtime files
              library = vim.api.nvim_get_runtime_file("", true),
            },
            -- Do not send telemetry data containing a randomized but unique identifier
            telemetry = {
              enable = false,
            },
          },
        },
      })
    end,
  },
```

~~@folke I'm not really sure this is the best way to achieve completions in the neovim/LazyVim lua config files. But is there a reason you wouldn't want something like this in LazyVim's lspconfig (eventhough it's specific to the `lua-language-server`)?~~

</details>

**Reply by jyuan0** [reply]
Maybe this is an issue with your setup? `gd` works out of the box for me, and I think those settings are already set by neodev et al.

**Reply by fredrikaverpil** [reply]
@jyuan0 so you have working `gd` and completions out of the box with LazyVim for files in the `lua/plugins` folder?

EDIT: I just temporarily moved all plugins away, and indeed, I can do `gd` on some functions... hm. I'll investigate and see what in my setup is borking this.

**Reply by jyuan0** [reply]
Good luck!

**Reply by fredrikaverpil** [reply]
@jyuan0 I don't really know what was wrong 😅 but after having moved all lua files out of the `plugin` folder and back again, I now have working `gd`. My (edited out) snippet above borked the `gd` too, so that wasn't good for much!

But just to be clear, you don't have the ability to hit `gd` on anything that is required, like the `mason_registry` in the below snippet, right? (this was what my initial post/question was about)

```lua
return {
  {
    "mason.nvim",
    config = function()
      local mason_registry = require("mason-registry")  -- `gd` does not work here at all for me
    end,
  },
}
```

**Reply by dpetka2001** [reply]
I have the following in my config for dap-python
```lua
      config = function()
        local path = require("mason-registry").get_package("debugpy"):get_install_path()
        require("dap-python").setup(path .. "/venv/bin/python")
      end,
```
and if I `gd` with the mouse pointer on `mason-registry` it takes me to `~/.local/share/nvim/lazy/mason.nvim/lua/mason-registry/init.lua
` where it is defined. Neodev should take care of that automatically.

**Reply by jyuan0** [reply]
> and if I gd with the mouse pointer on mason-registry it takes me to ~/.local/share/nvim/lazy/mason.nvim/lua/mason-registry/init.lua  where it is defined.

Same for me.

**Reply by fredrikaverpil** [reply]
Okay, I think I understand what's going on. I've completely erased and re-installed LazyVim and now I can hit `gd` on the "mason-registry" string in `requires("mason-registry")` just fine.

I used to install LazyVim a bit differently, so that I could easily try out different distros, and I believe this is why I initially was having issues here.

What I did was I took advantage of the `NVIM_APPNAME` environment variable. Meaning, I could install and run LazyVim like this:

```bash
# note that I'm NOT installing into ~/.config/nvim
git clone --recursive https://github.com/LazyVim/starter.git ~/.config/LazyVim

# to run LazyVim
NVIM_APPNAME=LazyVim nvim
```

This way I could install several distros and jump between them. I set up a simple installation which installs a few distros [here](https://github.com/fredrikaverpil/dotfiles/blob/main/installers/neovim.sh) along with a fzf picker [here](https://github.com/fredrikaverpil/dotfiles/blob/main/shell/bin/nvims). Then finally I aliased `nvim` to use my preferred choice [here](https://github.com/fredrikaverpil/dotfiles/blob/main/shell/aliases.sh#L44-L45).

This is all something I originally saw demoed by @elijahmanor 👏😄 here: https://www.youtube.com/watch?v=LkHjJlSgKZY&t=3s

But anyhow, I believe now since I'm configuring the different distros in locations that might not be expected by e.g. LazyVim, such as `~/.config/LazyVim/lua/plugins`, some stuff breaks, such as when attempting to hit `gd` on strings within `requires()` statements.

**Reply by fredrikaverpil** [reply]
@folke would it be possible for LazyVim to support `NVIM_APPNAME`, do you think?

**Reply by jyuan0** [reply]
Are you saying your issue is related to `NVIM_APPNAME` and not because of something in your config? `gd` doesn't work on `requires("mason-registry")` on a clean install of LazyVim if you install it the non-standard way?

**Reply by dpetka2001** [reply]
What do you mean for LazyVim to support `NVIM_APPNAME`? I also have different distros with the way you described and everything works correctly since they all use different locations for their settings. I have the default LazyVim installed in `.config/nvim` and then some others in different folders and invoke them through the fzf picker.

**Reply by fredrikaverpil** [reply]
> Are you saying your issue is related to NVIM_APPNAME and not because of something in your config?

Yes, it appears so.

> `gd` doesn't work on requires("mason-registry") on a clean install of LazyVim if you install it the non-standard way?

Yes. When I install the non-standard way, I can e.g. hit `gd` on `requires` but not on the `"mason-registry"` string.

**Reply by fredrikaverpil** [reply]
> What do you mean for LazyVim to support NVIM_APPNAME?

My guess is that whatever sets up `gd` to work with LazyVim's lua config files (neodev?) searches in certain locations but not in `~/.config/$NVIM_APPNAME/lua/plugins`. I'm not sure though.

> I also have different distros with the way you described and everything works correctly since they all use different locations for their settings. I have the default LazyVim installed in .config/nvim and then some others in different folders and invoke them through the fzf picker.

I guess the key for you here is you have your plugins in `~/.config/nvim/lua/plugins` but I don't. I have all my plugins in `~/.config/LazyVim/lua/plugins`. Maybe I should consider using the default `~/.config/nvim` location for LazyVim as this is my primary go-to distro...

**Reply by jyuan0** [reply]
You can find out what directories `lua_ls` has configured for the workspace with `Neoconf lsp`.

**Reply by fredrikaverpil** [reply]
@dpetka2001 sigh, so it might be something else. I just realized I do have `gd` support on the requires strings when using `$NVIM_APPNAME`. It might all have something to do with that I'm symlinking my lua config files into place from my dotfiles repo.

**Reply by dpetka2001** [reply]
I have a secondary distro for testing purposes in `~/.config/lazyTest` and invoke Neovim with `NVIM_APPNAME=lazyTest nvim`. My plugins for the testing distro reside in `~/.config/lazyTest/lua/plugins`. `gd` on `require("mason-registry")` behaves the same. `Neoconf lsp` shows that the paths in `lua_ls.Lua.workspace.library` contain the paths from the `lazyTest` instead of the default `nvim` configuration. They are not mixed up and everything works correctly in both seperate Neovim invocations with seperate configuration directories.

**Reply by fredrikaverpil** [reply]
@dpetka2001 thanks for the clarification. Are you using symlinks, like symlinking the `lua` folder into `~/.config/<folder>/lua`?

I can reliably reproduce here that as soon as I'm symlinking `~/code/repos/dotfiles/nvim-lazyvim/lua` to `~/.config/LazyVim/lua`, I can't hit `gd` on the `requires` string value when running nvim with `NVIM_APPNAME=LazyVim nvim`.

`Neoconf lsp` output with symlink active, and as you can see the workspace includes the files from my symlinked dotfiles repo:

```lua
# Lsp Settings

## lua_ls

* ~/code/repos/dotfiles/.vscode/settings.json
{
  Lua = {
    completion = {
      callSnippet = "Replace"
    },
    runtime = {
      path = { "?.lua", "?/init.lua" },
      pathStrict = true,
      version = "LuaJIT"
    },
    telemetry = {
      enable = false
    },
    workspace = {
      checkThirdParty = false,
      ignoreDir = { "types/nightly", "lua" },
      library = { "/Users/fredrik/.local/share/LazyVim/lazy/neodev.nvim/types/stable", "/opt/homebrew/Cellar/neovim/0.9.1/share/nvim/runtime/lua", "/Users/fredrik/code/repos/dotfiles/nvim-lazyvim/lua" }
    }
  },
  peacock = {
    color = "#f27983"
  },
  workbench = {
    colorCustomizations = {
      ["activityBar.activeBackground"] = "#f7a7ae",
      ["activityBar.activeBorder"] = "#19920d",
      ["activityBar.background"] = "#f7a7ae",
      ["activityBar.foreground"] = "#15202b",
      ["activityBar.inactiveForeground"] = "#15202b99",
      ["activityBarBadge.background"] = "#19920d",
      ["activityBarBadge.foreground"] = "#e7e7e7",
      ["commandCenter.border"] = "#15202b99",
      ["sash.hoverBorder"] = "#f7a7ae",
      ["statusBar.background"] = "#f27983",
      ["statusBar.foreground"] = "#15202b",
      ["statusBarItem.hoverBackground"] = "#ed4a58",
      ["statusBarItem.remoteBackground"] = "#f27983",
      ["statusBarItem.remoteForeground"] = "#15202b",
      ["titleBar.activeBackground"] = "#f27983",
      ["titleBar.activeForeground"] = "#15202b",
      ["titleBar.inactiveBackground"] = "#f2798399",
      ["titleBar.inactiveForeground"] = "#15202b99"
    }
  }
}

## null-ls

* ~/code/repos/dotfiles/.vscode/settings.json

{}
```


As soon as I stop using a symlink for the `lua` folder, and instead copy the files into ~/.config/LazyVim/lua/plugins`, I can successfully hit `gd` on the "mason-registry" string.

**Reply by jyuan0** [reply, marked as answer]
Would one of those directories happen to have a `.neoconf.json` file, and one not?

**Reply by fredrikaverpil** [reply]
@jyuan0 yes, there is a `~/.config/LazyVim/.neoconf.json`:
```json
{
  "neodev": {
    "library": {
      "enabled": true,
      "plugins": true
    }
  },
  "neoconf": {
    "plugins": {
      "lua_ls": {
        "enabled": true
      }
    }
  }
}
```

Copying this into my `~/code/repos/dotfiles/nvim-lazy` made `gd` finally work consistently! 🥳

**Reply by jyuan0** [reply]
`.neoconf.json` defines a local Neoconf workspace, and the `plugins: true` (under `neodev`) should add plugin directories to your `lua_ls` workspace library directories. See the docs for Neodev and Neoconf.

**Reply by jyuan0** [reply]
Maybe that's not your issue, but I thought best to be aware of this.

**Reply by dpetka2001** [reply]
Glad you figured it out. I don't do any symlinking myself. I use chezmoi to backup my files which doesn't use symlinking.

**Reply by fredrikaverpil** [reply]
@jyuan0 @dpetka2001 many, many thanks for helping me get to this point where it finally works. It helps so much to grok things when you can read the plugin code 😄

**Reply by jyuan0** [reply]
You're welcome. Glad you figured it out too.
